nkms/network/protocols.py

```diff
@@ -1,13 +1,43 @@
+from kademlia.node import Node
 from kademlia.protocol import KademliaProtocol
-import asyncio
+from nkms.network.routing import NuCypherRoutingTable
 
 
 class NuCypherHashProtocol(KademliaProtocol):
-    pass
+
+    def __init__(self, sourceNode, storage, ksize, *args, **kwargs):
+        super().__init__(sourceNode, storage, ksize, *args, **kwargs):
+        self.router = NuCypherRoutingTable(self, ksize, sourceNode)
+
+    def rpc_ping(self, sender, nodeid, node_properties={}):
+        source = Node(nodeid, sender[0], sender[1])
+        self.welcomeIfNewNode(source, node_properties)
+        return self.sourceNode.id
+
+    def welcomeIfNewNode(self, node, node_properties):
+        if not self.router.isNewNode(node):
+            return
+
+        self.log.info("never seen %s before, adding to router and setting nearby " % node)
+        for key, value in self.storage.items():
```

Comment:
I didn't realize we have to iterate over all storage. Although, perhaps, it is possible to do some sort of index here (e.g. each node stores rekeys ordered by distanceTo(keynode) because it's always the same for this node, and iterate over saved elements which have the largest distance (because they're candidates to go away)